#63461042 get one office endpoint

vic3king opened the pull request:
#### What does this PR do?
The application should have an endpoint to get a specific political office
#### Description of Task to be completed?
- add model and controller files
- make GET request to /offices/:id
- write unit tests for the endpoint
#### How should this be manually tested?
- clone repo
- npm run dev-start 
- make GET request to api/vi/offices/:id
#### Any background context you want to provide?
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/n/projects/2238799/stories/#163461042
https://www.pivotaltracker.com/n/projects/2238799/stories/#163465853
#### Screenshots (if appropriate)
![getonee](https://user-images.githubusercontent.com/40821284/51788071-b0a2ef80-2171-11e9-865d-223426a2fe96.PNG)
![getone](https://user-images.githubusercontent.com/40821284/51788072-b13b8600-2171-11e9-869d-6773754aa8bb.PNG)

#### Questions:

vic3king committed:
ft(endpoint /parties): create endpoint for creating a party

- add model and contoller files
- make POST request to /parties
- write unit tests for the endpoint

[delivers #163460933 #163465686]

vic3king committed:
bg(POST/parties): fix validation bugs

- trim whitespaces
- validate url logo
- improve tests

[fixes #163492036]

vic3king committed:
ft(GET/parties/id): user should be able to get a specific party by id

- add model and contoller files
- make GET request to /parties/:id
- write unit tests for the endpoint

[delivers #163460972 #163465693]

vic3king committed:
fix failing test

vic3king committed:
ft(get all parties): user can get all existing parties

- add model and contoller files
- make GET request to /parties
- write unit tests for the endpoint

[delivers #163460951 #163465722]

vic3king committed:
feature(partyname): users can update a party name

- add model and contoller files
- make PATCH request to /parties/:id/name
- write unit tests for the endpoint

[delivers #163460983 #163465743]

vic3king committed:
feature(delete party): users should be able to delete a party

- add model and contoller files
- make PATCH request to /parties/:id/name
- write unit tests for the endpoint

[delivers #163460997 #163465775]

vic3king committed:
feature(create office): users hsould be able to create a political party

- add model and controller files
- make POST request to /offices
- write unit test for endpoint

[delivers #163461006 #163465804]

vic3king committed:
feature(get offices): there should be an endpoint that retrieves all existing offices

-- add model and contoller files
- make GET request to /offices
- write unit tests for the endpoint

[delivers #163461027 #163465819]

vic3king committed:
feature(delete party): users should be able to delete a party  …

- add model and controller files
- make GET request to /offices/:id
- write unit tests for the endpoint

+[delivers #63461042 #163465853]

JCanaks commented on .env:
```diff
@@ -0,0 +1 @@
+GOOGLE=AIzaSyCuprhlOtAFpOfhaYMs5fYdjdnnla57BLg
```
resolve EOL

JCanaks commented on src/controllers/office.js:
```diff
@@ -0,0 +1,55 @@
+const ControllerOffice = {
+  createOffice(req, res) {
+    // eslint-disable-next-line no-restricted-globals
+    if (isNaN(req.body.age)) {
```
@vic3king what is the concept behind the age attribute for an office? Also pls follow the format specified for the response `id`, `type`, `name` (not officeName)
<img width="634" alt="screen shot 2019-01-28 at 6 51 05 pm" src="https://user-images.githubusercontent.com/26818808/51855495-b47b7100-232d-11e9-85a5-f70eccd89a7d.png">

JCanaks replied:
Ok upon futher review understand the age concept. Nice addition. Pls take note of the response format specified

JCanaks replied:
To enable your code to be more readable you can use `ageLimit` instead of just `age` i believe that will be more approrpiate

JCanaks commented on src/controllers/office.js:
```diff
@@ -0,0 +1,55 @@
+    // eslint-disable-next-line no-restricted-globals
+    if (isNaN(req.body.age)) {
+      return res.status(406).json({
+        status: 406,
```
Error message should be more descriptive. What must be a number? message should say something like `Age Limit entered must be a number`

JCanaks commented on src/middlewares/helperoffice.js:
```diff
@@ -0,0 +1,70 @@
+const validId = id => Number.isInteger(parseInt(id, 10));
+const ValidateOffice = {
+  spaces(obj) {
+    const strType = obj.type.split(' ').join('');
```
@vic3king This whole block of code for checking spaces could have just been simplified by using the trim() function. so instead of `obj.type.split(' ').join('');` you just do a `obj.type.trim();` if it is only empty spaces it will become an empty string `""` so checking the length will be less than 1 or you can just check if it is defined --> `if(strType){.....}` an empty string is undefined (no need to check length)

JCanaks commented on src/controllers/party.js:
```diff
@@ -0,0 +1,79 @@
+    }
+    const address = request.hqAddressUrl;
+    let {
+      id, name, hqAddressUrl, logoUrl, status, createdOn, createdBy,
```
Following the format on the ADC docs a party should have an `hqAddress` representative of the address of the party's headquarters not `hqAddressUrl`. Take note pls
<img width="544" alt="screen shot 2019-01-29 at 12 59 59 am" src="https://user-images.githubusercontent.com/26818808/51874497-7c8e2100-2361-11e9-8d78-0d158537f10e.png">

JCanaks commented on src/middlewares/helper.js:
```diff
@@ -0,0 +1,119 @@
+        if (res.json.results) {
+          request.hqAddressUrl = res.json.results[0].formatted_address;
+        } else {
+          return response.status(404).send({
```
this line of code if unreachable when there is no result from the Google API `res.json.results` becomes an empty array --> `[]` checking if an empty array is defined (if (res.json.results) {...}) returns true so the else case is never captured and instead an error about formatted address pops up. its better to use  `if (res.json.results.length < 1) {}` this will capture the cases where there are no results

vic3king replied:
thank you for the feedback, i will implement them.

JCanaks reviewed (commented):
@vic3king Nice work but take note of all feedback

JCanaks commented on src/test/test.js:
```diff
@@ -34,3 +34,560 @@ describe('UI routes', () => {
 });
+
+
+describe('/Post create political party', () => {
```
Separate test into different files according to what you are testing for. So all tests for party should be in a file called `party.test.js` or just `party.js` and so on. That way your code is cleaner and failing tests can easily be identified instead of having a bulk load of tests in one file

vic3king committed:
ft(GET/offices/id): user should be able to get a specific party by id

- add model and contoller files
- make GET request to /offices/:id
- write unit tests for the endpoint

[delivers #63461042]

vic3king committed:
implemet LFA feedback

vic3king committed:
Merge branch 'ft-get-one-office-#63461042' of https://github.com/vic3king/politico into ft-get-one-office-#63461042
